lib/Sema/BuilderTransform.cpp

```diff
@@ -80,10 +80,21 @@ class BuilderClosureVisitor
     auto simplifiedTy = cs->simplifyType(builderType);
     if (!simplifiedTy->hasTypeVariable()) {
       typeExpr = TypeExpr::createImplicitHack(loc, simplifiedTy, ctx);
+    } else if (auto *decl = simplifiedTy->getAnyGeneric()) {
+      // HACK: If there's not enough information to completely resolve the
+      // builder type, but we have the base available to us, form an *explicit*
+      // TypeExpr pointing at it. We cannot form an implicit base without
+      // a fully-resolved concrete type. Really, whatever we put here has no
+      // bearing on the generated solution because we're going to use this node
+      // to stash the builder type and hand it back to the ambient
+      // constraint system.
+      typeExpr = TypeExpr::createForDecl(DeclNameLoc(loc), decl, dc);
```

Comment:
@CodaFi  I agree with @hamishknight, it seems like we need to somehow connect type variables which are going to be generated for generic arguments of a new call (after base would be re-opened) to type variables representing the same generic arguments in the outer scope...

Reply:
Right. The old patch used to do this by peeling the type out of the TypeLoc (and trying to re-open it! https://github.com/apple/swift/commit/09db2902d21107165225f3397ac45e0e7c9fb018#diff-ba9f048d1813343154c6702f3188c8b8L1466 😱).

Reply:
@CodaFi Wouldn't it have hit the `CS.getType(typeLoc)` code path? `wasValidated()` should be `false` for a null `TypeLoc`.

Reply:
Yeah, that's the hack we need to put back here.